src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/typeresolution.cpp

```diff
@@ -0,0 +1,657 @@
+            hr = GetNewMetaDataImporter(
+                pMetaDataDispenser,
+                pszCandidateFilePath,
+                ppMetaDataImporter);
+        }
+
+        LeaveCriticalSection(&_csCacheLock);
+
+        return hr;
+    }
+
+    HRESULT MetaDataImportersLRUCache::GetNewMetaDataImporter(
+        _In_ IMetaDataDispenserEx* pMetaDataDispenser,
+        _In_ PCWSTR pszCandidateFilePath,
+        _Outptr_opt_ IMetaDataImport2** ppMetaDataImporter)
+    {
+        if (ppMetaDataImporter == nullptr)
+        {
+            return ERROR_BAD_ARGUMENTS;
+        }
+
+        HRESULT hr;
+
+        hr = pMetaDataDispenser->OpenScope(
+            pszCandidateFilePath,
+            ofReadOnly,
+            IID_IMetaDataImport2,
+            reinterpret_cast<IUnknown**>(ppMetaDataImporter));
+
+        if (SUCCEEDED(hr))
+        {
+           _metadataImportersMap.emplace(
+                pszCandidateFilePath,
+                *ppMetaDataImporter);
+        }
+
+        if (SUCCEEDED(hr))
+        {
+            hr = AddNewFilePathToList(pszCandidateFilePath);
+        }
+
+        return hr;
+    }
+
+    HRESULT MetaDataImportersLRUCache::AddNewFilePathToList(PCWSTR pszFilePath)
+    {
+        HRESULT hr = RemoveLeastRecentlyUsedItemIfListIsFull();
+
+        if (SUCCEEDED(hr))
+        {
+            // Make room for new element.
+            for (int i = g_dwMetaDataImportersLRUCacheSize - 2; i >= 0; i--)
+            {
+                _arFilePaths[i + 1] = _arFilePaths[i];
+            }
+
+            _arFilePaths[0] = AllocateAndCopyString(pszFilePath);
+
+            if (_arFilePaths[0] == nullptr)
+            {
+                hr = E_OUTOFMEMORY;
+            }
+        }
+
+        return hr;
+    }
+
+    bool MetaDataImportersLRUCache::IsFilePathCached(PCWSTR pszFilePath)
+    {
+        int filePathIndex = GetFilePathIndex(pszFilePath);
+
+        if (filePathIndex != -1)
+        {
+            MoveElementToFrontOfList(filePathIndex);
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    int MetaDataImportersLRUCache::GetFilePathIndex(PCWSTR pszFilePath)
+    {
+        int filePathIndex = -1;
+
+        for (int i = 0; (i < g_dwMetaDataImportersLRUCacheSize) && (_arFilePaths[i] != nullptr); i++)
+        {
+            if (wcscmp(pszFilePath, _arFilePaths[i]) == 0)
+            {
+                filePathIndex = i;
+                break;
+            }
+        }
+
+        return filePathIndex;
+    }
+
+    void MetaDataImportersLRUCache::MoveElementToFrontOfList(int elementIndex)
+    {
+        PWSTR pszFoundFilePath = _arFilePaths[elementIndex];
+
+        for (int i = elementIndex - 1; i >= 0; i--)
+        {
+            _arFilePaths[i + 1] = _arFilePaths[i];
+        }
+
+        _arFilePaths[0] = pszFoundFilePath;
+    }
+
+    HRESULT MetaDataImportersLRUCache::RemoveLeastRecentlyUsedItemIfListIsFull()
```

Comment:
If the list is full, will we not check all metadata?

Reply:
No it will right? This is basically guaranteeing for AddNewFilePathToList that it can add a new file path and the list won't be full.

Reply:
Couldn't it in theory add enough new files that the first one get's pushed out?  Meaning we wouldn't then search that file?  Is this even a scenario of concern?  I'm not sure how many files the list can hold, but it's likely plenty.

Reply:
It should only hold up to g_dwMetaDataImportersLRUCacheSize. It's based on least recently used (LRU). So in theory, GetMetaDataImporter should be removing the least used one if the size is full. But if it's get called again on the winmd that was removed, it should readd it to the cache by calling GetNewMetaDataImporter on it.